### PR TITLE
MBS-9135: Allow admins to modify usernames

### PR DESCRIPTION
**Pull request by reosarevok**
Tested on local sandbox and it seems to work just fine, but let me know if something seems off!

### CODE REVIEW

**Comment by Freso** [comment]
As the user would need to log out and back in again for the change to "show up", is there any chance that all their current sessions could get removed/invalidated once the name is changed? (I looked over the code and it doesn't look like this is done currently.)

File: lib/MusicBrainz/Server/Form/Role/ValidateUsername.pm [hunk 1]
```diff
@@ -0,0 +1,22 @@
...
+use MusicBrainz::Server::Translation qw( l );
+
+sub validate_username
+{
+    my ($self, $field) = @_;
+
+    my $username = $field->value;
+    if (defined $username) {
+        if ($username =~ qr{^deleted editor \#\d+$}i) {
+            $field->add_error(l('This username is reserved for internal use'));
```
Comment (Freso) [inline lib/MusicBrainz/Server/Form/Role/ValidateUsername.pm]: Are these the only names reserved for internal use? Also, do we not use proper grammar (ie., punctuation) in other messages?

File: lib/MusicBrainz/Server/Form/Role/ValidateUsername.pm [hunk 2]
```diff
@@ -0,0 +1,22 @@
...
+        }
+        my $editor = $self->ctx->model('Editor')->get_by_name($username);
+        if (defined $editor) {
+            $field->add_error(l('Please choose another username, this one is already taken'));
```
Comment (Freso) [inline lib/MusicBrainz/Server/Form/Role/ValidateUsername.pm]: Punctuation?

File: lib/MusicBrainz/Server/Form/Role/ValidateUsername.pm (hunk 1 above)
Comment (reosarevok) [inline lib/MusicBrainz/Server/Form/Role/ValidateUsername.pm]: I'd expect they are (this is the validation we were already doing for registering, I've just moved it to a role). Not sure about punctuation, https://github.com/metabrainz/musicbrainz-server/search?q=add_error&type=Code&utf8=%E2%9C%93 suggests the answer is "sometimes"

Comment (Freso) [inline lib/MusicBrainz/Server/Form/Role/ValidateUsername.pm]: Well, if it's "sometimes", let's move towards more consistently using proper grammar than now. :)

File: lib/MusicBrainz/Server/Form/Role/ValidateUsername.pm [hunk 3]
```diff
@@ -0,0 +1,22 @@
+package MusicBrainz::Server::Form::Role::ValidateUsername;
```
Comment (chirlu) [inline lib/MusicBrainz/Server/Form/Role/ValidateUsername.pm]: I don’t think this is appropriate to have as a role. Make it a helper function somewhere.

File: lib/MusicBrainz/Server/Form/User/EditProfile.pm [hunk 4]
```diff
@@ -7,9 +7,15 @@ use MusicBrainz::Server::Translation qw( l ln );
 use MusicBrainz::Server::Validation qw( is_valid_url );
 
 extends 'MusicBrainz::Server::Form';
+with 'MusicBrainz::Server::Form::Role::ValidateUsername';
 
 has '+name' => ( default => 'profile' );
 
+has_field 'username' => (
+    type      => 'Text',
+    maxlength => 64
```
Comment (chirlu) [inline lib/MusicBrainz/Server/Form/User/EditProfile.pm]: Should have a trailing comma, so that adding more entries later doesn’t require changing this line.

File: lib/MusicBrainz/Server/Controller/Admin.pm [hunk 5]
```diff
@@ -35,22 +35,22 @@ sub edit_user : Path('/admin/user/edit') Args(1) RequireAuth HiddenOnSlaves
...
         form => 'User::EditProfile',
         item => {
-            email            => $user->email,
+            username          => $user->name,
+            email             => $user->email,
             skip_verification => 0,
-            website            => $user->website,
-            biography        => $user->biography
+            website           => $user->website,
+            biography         => $user->biography
```
Comment (chirlu) [inline lib/MusicBrainz/Server/Controller/Admin.pm]: I wouldn’t bother with this optical alignment stuff. If you do, make it a separate commit please.

File: lib/MusicBrainz/Server/Controller/Admin.pm [hunk 6]
```diff
@@ -35,22 +35,22 @@ sub edit_user : Path('/admin/user/edit') Args(1) RequireAuth HiddenOnSlaves
...
         },
     );
 
     if ($c->form_posted) {
-        if ($form->submitted_and_valid($c->req->params )) {
+        if ($form->submitted_and_valid($c->req->params) 
+            && $form2->submitted_and_valid($c->req->params )) {
```
Comment (chirlu) [inline lib/MusicBrainz/Server/Controller/Admin.pm]: Combining the `if`s changes the semantics, unless both forms will always be submitted at the same time. Are you sure that is the case? Why?

File: lib/MusicBrainz/Server/Form/User/EditProfile.pm [hunk 7]
```diff
@@ -7,9 +7,15 @@ use MusicBrainz::Server::Translation qw( l ln );
 use MusicBrainz::Server::Validation qw( is_valid_url );
 
 extends 'MusicBrainz::Server::Form';
+with 'MusicBrainz::Server::Form::Role::ValidateUsername';
 
 has '+name' => ( default => 'profile' );
 
+has_field 'username' => (
```
Comment (chirlu) [inline lib/MusicBrainz/Server/Form/User/EditProfile.pm]: I expect this will let everyone change their own username. (If not, how/where is it prevented?)

File: lib/MusicBrainz/Server/Controller/Admin.pm (hunk 6 above)
Comment (reosarevok) [inline lib/MusicBrainz/Server/Controller/Admin.pm]: As far as I can tell, both forms *should* always be submitted at the same time. There's only one form element, with one submit button, pressing Enter anywhere submits both forms from what I've tested, and the separation into two forms seems to exist only in the FormHandler part of this. But I'm not super experienced with this, so I might be missing something. Am I?

File: lib/MusicBrainz/Server/Form/User/EditProfile.pm (hunk 7 above)
Comment (reosarevok) [inline lib/MusicBrainz/Server/Form/User/EditProfile.pm]: It's not, specifically, right now, other than by not having a username field in the template on the non-Admin page. Are you thinking we need to implement an extra check in case someone tries to change anyway it by post request or something to that effect?

File: lib/MusicBrainz/Server/Form/Role/ValidateUsername.pm (hunk 3 above)
Comment (reosarevok) [inline lib/MusicBrainz/Server/Form/Role/ValidateUsername.pm]: Any suggestions as to where this "somewhere" should be?

Comment (mwiencek) [inline lib/MusicBrainz/Server/Form/Role/ValidateUsername.pm]: I think a role could make sense for username behavior in general, so also adding `has_field 'username'` here at least, assuming the `required` attribute works in both places. But, just moving the function to lib/MusicBrainz/Server/Validation.pm is sane too.

Comment (mwiencek) [inline lib/MusicBrainz/Server/Form/Role/ValidateUsername.pm]: Or maybe lib/MusicBrainz/Server/Form/Utils.pm instead, since the Validation module doesn't really contain anything that operates on FormHandler fields directly right now. Having a generic "Form::Role::Username" still makes sense too IMHO.

File: lib/MusicBrainz/Server/Form/Role/ValidateUsername.pm [hunk 8]
```diff
@@ -0,0 +1,24 @@
...
+use HTML::FormHandler::Moose::Role;
+use MusicBrainz::Server::Translation qw( l );
+
+sub validate_username
+{
+    my ($self, $field) = @_;
+
+    my $username = $field->value;
+    if (defined $username) {
+        unless ($self->ctx->user->name == $username) {
```
Comment (mwiencek) [inline lib/MusicBrainz/Server/Form/Role/ValidateUsername.pm]: The `validate_username` function that was in the Register form didn't have this check...does account creation still work? I expect it would either ISE or not run the validation.

File: lib/MusicBrainz/Server/Form/User/EditProfile.pm (hunk 7 above)
Comment (mwiencek) [inline lib/MusicBrainz/Server/Form/User/EditProfile.pm]: We'll need a check for that in `Controller::Account`, I think. I would just `$c->detach('/error_403')` if the username changed and the user isn't an account admin.

File: lib/MusicBrainz/Server/Controller/Admin.pm [hunk 9]
```diff
@@ -76,11 +76,11 @@ sub edit_user : Path('/admin/user/edit') Args(1) RequireAuth HiddenOnSlaves
                     $c->forward('/discourse/sync_sso', [$user]);
                 }
             }
+            
+	        $c->flash->{message} = l('User successfully edited.');
```
Comment (mwiencek) [inline lib/MusicBrainz/Server/Controller/Admin.pm]: There are some tab characters at the beginning of these 3 lines.

File: lib/MusicBrainz/Server/Controller/Admin.pm (hunk 6 above)
Comment (mwiencek) [inline lib/MusicBrainz/Server/Controller/Admin.pm]: That sounds correct, i.e. they should both be submitted at the same time. It looks like it was previously the case that you'd get the redirect and "User successfully edited" banner even if one of the two forms contained errors.

File: lib/MusicBrainz/Server/Controller/Account.pm [hunk 10]
```diff
@@ -297,6 +297,13 @@ sub edit : Local RequireAuth DenyWhenReadonly {
             $form->value
         );
 
+        my $old_username = $editor->username;
+        my $new_username = $form->field('username')->value;
+              
+        if (defined $new_username && $new_username ne $old_username) {
+            $c->detach('/error_403');            
+        }
+
```
Comment (reosarevok) [inline lib/MusicBrainz/Server/Controller/Account.pm]: Can't test this, because the account/edit form seems unrelatedly bugged (about languages) but thought I'd put it up so you could say if it looks reasonable or not at least.

File: lib/MusicBrainz/Server/Controller/Account.pm [hunk 11]
```diff
@@ -297,6 +297,13 @@ sub edit : Local RequireAuth DenyWhenReadonly {
             $form->value
         );
 
+        my $old_username = $editor->username;
```
Comment (mwiencek) [inline lib/MusicBrainz/Server/Controller/Account.pm]: That should be `->name` (although this code doesn't actually run yet, due to an error below).

File: lib/MusicBrainz/Server/Form/User/EditProfile.pm [hunk 12]
```diff
@@ -2,14 +2,21 @@ package MusicBrainz::Server::Form::User::EditProfile;
...
 use MusicBrainz::Server::Translation qw( l ln );
 use MusicBrainz::Server::Validation qw( is_valid_url );
 
 extends 'MusicBrainz::Server::Form';
 
 has '+name' => ( default => 'profile' );
 
+has_field 'username' => (
+    type      => 'Text',
+    required  => 1,
```
Comment (mwiencek) [inline lib/MusicBrainz/Server/Form/User/EditProfile.pm]: It looks like the `required` field actually breaks the profile page, because since there's no username field on the page, nothing is submitted and an invisible error occurs.

File: lib/MusicBrainz/Server/Form/User/EditProfile.pm [hunk 13]
```diff
@@ -2,14 +2,21 @@ package MusicBrainz::Server::Form::User::EditProfile;
 
 use HTML::FormHandler::Moose;
 use List::MoreUtils qw( any all );
-use MusicBrainz::Server::Form::Utils qw( language_options select_options_tree );
+use MusicBrainz::Server::Form::Utils qw( language_options select_options_tree validate_username);
```
Comment (mwiencek) [inline lib/MusicBrainz/Server/Form/User/EditProfile.pm]: Nitpick: space is missing before the closing `)` like the rest of the imports (same problem in lib/MusicBrainz/Server/Form/User/Register.pm).

File: lib/MusicBrainz/Server/Form/User/EditProfile.pm (hunk 12 above)
Comment (mwiencek) [inline lib/MusicBrainz/Server/Form/User/EditProfile.pm]: Since `required` is important on admin/edit_user, a simple fix would be to add an `<input type="hidden />` containing the username on the profile page.

File: lib/MusicBrainz/Server/Form/Utils.pm [hunk 14]
```diff
@@ -231,6 +232,25 @@ sub indentation {
     return "\N{NO-BREAK SPACE}" x (3 * $level);
 }
 
+sub validate_username {
+    my ($self) = @_;
+
+    my $username = $self->value;
+    my $previous_username = $self->init_value;
+
+    if (defined $username) {
```
Comment (mwiencek) [inline lib/MusicBrainz/Server/Form/Utils.pm]: I think `$username` should always be defined, since it's marked as `required` in both forms?

Comment (reosarevok) [inline lib/MusicBrainz/Server/Form/Utils.pm]: That does sound reasonable, yeah... Wonder why the check was here originally then, given this was only used in the registration page!

Comment (mwiencek) [inline lib/MusicBrainz/Server/Form/Utils.pm]: Hm, don't trust me too much on that, it's possible it runs all the validation handlers even if it fails the `required` one.

**Comment by gentlecat** [comment]
Not directly related to this PR, but I'm pretty sure that username is used as a reference to a user in all the other projects (ours) that rely on OAuth. So if username changes it will affect ability to access accounts in AB, CB, etc. In the worst case, if a MusicBrainz user gets username of an old account, they'll have access to it in other projects.

Don't remember if there is a user ID or some other permanent reference that can be used as an alternative (and is used in OAuth).

**Comment by mwiencek** [comment]
Yes, we set `sub` (which is supposed to never be reassigned) to the username, which is pretty bad, but we can't change it without breaking every OAuth client. We do also provide the [user ID in the userinfo endpoint](https://github.com/metabrainz/musicbrainz-server/blob/b138e0d1fc5710e14ceee8f8e388f3894411683f/lib/MusicBrainz/Server/Controller/OAuth2.pm#L359-L359), but that's not a standard field. Preventing usernames from being reused in MBS would be a schema change.

File: lib/MusicBrainz/Server/Controller/Account.pm [hunk 15]
```diff
@@ -289,14 +289,35 @@ sub edit : Local RequireAuth DenyWhenReadonly {
...
         $c->model('Editor')->update_profile(
             $editor,
             $form->value
         );
 
+        my $old_username = $editor->name;
+        my $new_username = $form->field('username')->value;
+              
+        if (defined $new_username && $new_username ne $old_username) {
+            $c->detach('/error_403');
```
Comment (mwiencek) [inline lib/MusicBrainz/Server/Controller/Account.pm]: Needs to come before the `update_profile` call. :)

File: lib/MusicBrainz/Server/Form/User/Register.pm [hunk 16]
```diff
@@ -10,7 +11,8 @@ has '+name' => ( default => 'register' );
 has_field 'username' => (
     type      => 'Text',
     required  => 1,
-    maxlength => 64
+    maxlength => 64,
+    validate_method => \&validate_username
```
Comment (chirlu) [inline lib/MusicBrainz/Server/Form/User/Register.pm]: Trailing comma.

Comment (reosarevok) [inline lib/MusicBrainz/Server/Form/User/Register.pm]: Nothing else in this file has the trailing comma, so I didn't add it because then I'd have to add it everywhere and then it wouldn't belong in this PR anymore I guess...

File: lib/MusicBrainz/Server/Controller/Account.pm [hunk 17]
```diff
@@ -289,14 +289,35 @@ sub edit : Local RequireAuth DenyWhenReadonly {
...
+            username          => $editor->name,
+            email             => $editor->email,
+            skip_verification => 0,
+            website           => $editor->website,
+            biography         => $editor->biography,
+            gender_id         => $editor->gender_id,
+            area_id           => $editor->area_id,
+            area              => $editor->area,
+            birth_date        => $editor->birth_date,
+            languages         => $editor->languages
```
Comment (chirlu) [inline lib/MusicBrainz/Server/Controller/Account.pm]: Trailing comma.

File: lib/MusicBrainz/Server/Form/User/EditProfile.pm (hunk 12 above)
Comment (chirlu) [inline lib/MusicBrainz/Server/Form/User/EditProfile.pm]: Works of course, but is very ugly IMO.

Comment (mwiencek) [inline lib/MusicBrainz/Server/Form/User/EditProfile.pm]: My other idea was to split `Form::User::EditProfile`, with common functionality in a role.

File: lib/MusicBrainz/Server/Form/User/Register.pm (hunk 16 above)
Comment (mwiencek) [inline lib/MusicBrainz/Server/Form/User/Register.pm]: It's def. okay to add the trailing commas in new lines or lines you're changing anyway, even if other parts of the file don't use it. But, it's not like I'll reject a PR just 'cause it didn't have trailing commas*—it's just better to add them.

*If we checked code style in the tests using e.g. Perl::Critic, it'd be much easier to say "here, fix this list of issues if you want the build to pass," than having to make tedious comments on GitHub which usually aren't worth my time (and not because I don't care).
